# Solving Coin Combination for \$1
Consider a combinatorial problem of breaking \$1 U.S. into all possible combinations of coins. For instance, four quarters, two quarters and five 10 cent dimes, and so on. Now, if we have to add coins up to $1, we can use at most $1 coin, two 50 cent coins, four quarters, and so on. We can also mix different values. You can already see how we're going to need several nested loops to figure out all the combinations.

In [3]:
# importing the standard set of Python modules
import math
import collections
import dataclasses
import datetime
import numpy as np
import pandas as pd
import matplotlib.pyplot as pp  

In [4]:
combinations = []

for count_100 in range(1+1):
    for count_50 in range(2+1):
        for count_25 in range(4+1):
            for count_10 in range(10+1):
                for count_5 in range(20+1):
                    for count_1 in range(100+1):
                        if 100*count_100 + 50*count_50 + 25*count_25 + 10*count_10 + 5*count_5 + count_1 == 100:
                            combinations.append([count_100, count_50, count_25, count_10, count_5, count_1])

In [5]:
combinations

[[0, 0, 0, 0, 0, 100],
 [0, 0, 0, 0, 1, 95],
 [0, 0, 0, 0, 2, 90],
 [0, 0, 0, 0, 3, 85],
 [0, 0, 0, 0, 4, 80],
 [0, 0, 0, 0, 5, 75],
 [0, 0, 0, 0, 6, 70],
 [0, 0, 0, 0, 7, 65],
 [0, 0, 0, 0, 8, 60],
 [0, 0, 0, 0, 9, 55],
 [0, 0, 0, 0, 10, 50],
 [0, 0, 0, 0, 11, 45],
 [0, 0, 0, 0, 12, 40],
 [0, 0, 0, 0, 13, 35],
 [0, 0, 0, 0, 14, 30],
 [0, 0, 0, 0, 15, 25],
 [0, 0, 0, 0, 16, 20],
 [0, 0, 0, 0, 17, 15],
 [0, 0, 0, 0, 18, 10],
 [0, 0, 0, 0, 19, 5],
 [0, 0, 0, 0, 20, 0],
 [0, 0, 0, 1, 0, 90],
 [0, 0, 0, 1, 1, 85],
 [0, 0, 0, 1, 2, 80],
 [0, 0, 0, 1, 3, 75],
 [0, 0, 0, 1, 4, 70],
 [0, 0, 0, 1, 5, 65],
 [0, 0, 0, 1, 6, 60],
 [0, 0, 0, 1, 7, 55],
 [0, 0, 0, 1, 8, 50],
 [0, 0, 0, 1, 9, 45],
 [0, 0, 0, 1, 10, 40],
 [0, 0, 0, 1, 11, 35],
 [0, 0, 0, 1, 12, 30],
 [0, 0, 0, 1, 13, 25],
 [0, 0, 0, 1, 14, 20],
 [0, 0, 0, 1, 15, 15],
 [0, 0, 0, 1, 16, 10],
 [0, 0, 0, 1, 17, 5],
 [0, 0, 0, 1, 18, 0],
 [0, 0, 0, 2, 0, 80],
 [0, 0, 0, 2, 1, 75],
 [0, 0, 0, 2, 2, 70],
 [0, 0, 0, 2, 3, 65],
 [0, 0, 0, 2, 4

In [6]:
len(combinations)

293

It turns out that there are 293 ways to get $1 out of change. 

This is the solution to the problem. 
However, in this analysis, it often happens that the solution raises a new set of questions. 
 - How many ways to make $2 out of change, or three? 
 - Does the number of combination increase linearly or quadratically?

What we need to do is to take the code we wrote and generalize it to answer those questions. Before we do so, we make a couple of changes to our code to make it faster and easier to generalize. This is an example of **refactoring**. 

## Refactoring
The first change is that we will not loop by count, but by value using the step argument. For instance, instead of the loop over quarters from zero through four, we will loop over the values.

The second change is to recognize that we don't actually need the innermost loop. As long as the total up to that point is less or equal $1, then we can always make up the difference with pennies. So we write one fewer loop. This looks all right.